lager/detail/smart_lens.hpp

```diff
@@ -46,9 +58,32 @@ struct smart_lens_base
  * Try to find a nice lens type for the give type.
  */
 template <typename T>
-struct smart_lens : smart_lens_base
+struct smart_lens : smart_lens_base<T>
 {};
 
+struct functor_wrapper_t {
+    template <typename T>
+    const_functor<T> operator()(T&& t) const noexcept;
+};
+
+template <typename Lens, typename T>
+using viewed_t = decltype(
+    std::declval<Lens>()(functor_wrapper_t{})(std::declval<T>()).value);
+
+template <typename T>
+struct smart_lens<std::optional<T>>
+{
+    template <typename U>
+    static auto make(U &&u) {
+        // don't lift lenses that can already handle optionals
+        if constexpr (zug::meta::is_detected<viewed_t, U, std::optional<T>>::value) {
```

Comment:
This doesn't work quite right yet.
It detects if the given type is a lens, but not if it can handle optionals specifically...

Reply:
yeah I don't think I can SFINAE viewing the lens right because [return type deduction doesn't work with SFINAE](https://en.cppreference.com/w/cpp/language/function#Return_type_deduction) and pretty much all the lenses use return type deduction _everywhere_...

I don't know how to make this work right. I might have to just accept that `lens` in `cursor[lens]` has to handle optionals explicitly.

Reply:
AFAIK, SFINAE works with type deduction if you use `decltype(auto)` I believe, maybe we just need to tweak the lenses to use that in the right places.

Reply:
if we do decide to make all our API SFINAE friendly to enable more metaprogramming stuff, it means we need to enforce explicit return types on lambdas passed to things like `lens::getset`.